File: Lua/bucket_sort.lua
function bucket_sort(arr, bucket_size)
    bucket_size = bucket_size or 5
    local min_value, max_value = math.min(table.unpack(arr)), math.max(table.unpack(arr))

    local bucket_count = math.floor((max_value - min_value) / bucket_size) + 1
    local buckets = {}

    for i = 1, bucket_count do buckets[i] = {} end

    for _, v in ipairs(arr) do
        local index = math.floor((v - min_value) / bucket_size) + 1
        table.insert(buckets[index], v)
    end

    for _, bucket in ipairs(buckets) do
        table.sort(bucket)
    end

    local sorted_arr = {}
    for _, bucket in ipairs(buckets) do
        for _, v in ipairs(bucket) do
            table.insert(sorted_arr, v)
        end
    end

    return sorted_arr
end

local arr = {29, 25, 3, 49, 9, 37, 21, 43}
arr = bucket_sort(arr)
print(table.concat(arr, " "))

File: JavaScript/bucket_sort.js
function bucketSort(arr, bucketSize = 5) {
    if (arr.length === 0) {
        return arr;
    }

    let minValue = arr[0];
    let maxValue = arr[0];
    for (let i = 1; i < arr.length; i++) {
        if (arr[i] < minValue) {
            minValue = arr[i];
        } else if (arr[i] > maxValue) {
            maxValue = arr[i];
        }
    }

    let bucketCount = Math.floor((maxValue - minValue) / bucketSize) + 1;
    let buckets = new Array(bucketCount);
    for (let i = 0; i < buckets.length; i++) {
        buckets[i] = [];
    }

    for (let i = 0; i < arr.length; i++) {
        buckets[Math.floor((arr[i] - minValue) / bucketSize)].push(arr[i]);
    }

    arr.length = 0;
    for (let i = 0; i < buckets.length; i++) {
        buckets[i].sort((a, b) => a - b);
        arr.push(...buckets[i]);
    }

    return arr;
}

const array = [29, 25, 3, 49, 9, 37, 21, 43, 0];
console.log('Sorted array:', bucketSort(array));